source/velocity_boundary_conditions/ascii_data.cc

```diff
@@ -0,0 +1,439 @@
+                             "in the input file, this number is either interpreted as seconds or as years.");
+          prm.declare_entry ("First data file number", "0",
+                             Patterns::Integer (),
+                             "Number of the first velocity file to be loaded when the model time "
+                             "is larger than 'First velocity file model time'.");
+          prm.declare_entry ("Decreasing file order", "false",
+                             Patterns::Bool (),
+                             "In some cases the boundary files are not numbered in increasing "
+                             "but in decreasing order (e.g. 'Ma BP'). If this flag is set to "
+                             "'True' the plugin will first load the file with the number "
+                             "'First velocity file number' and decrease the file number during "
+                             "the model run.");
+          prm.declare_entry ("Scale factor", "1",
+                             Patterns::Double (0),
+                             "Scalar factor, which is applied to the boundary velocity. "
+                             "You might want to use this to scale the velocities to a "
+                             "reference model (e.g. with free-slip boundary) or another "
+                             "plate reconstruction. Another way to use this factor is to "
+                             "convert units of the input files. The unit is assumed to be"
+                             "m/s or m/yr depending on the 'Use years in output instead of "
+                             "seconds' flag. If you provide velocities in cm/yr set this "
+                             "factor to 0.01.");
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+
+    template <int dim>
+    void
+    AsciiData<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Boundary velocity model");
+      {
+        prm.enter_subsection("Ascii data model");
+        {
+          // Get the path to the data files. If it contains a reference
+          // to $ASPECT_SOURCE_DIR, replace it by what CMake has given us
+          // as a #define
+          data_directory    = prm.get ("Data directory");
+          {
+            const std::string      subst_text = "$ASPECT_SOURCE_DIR";
+            std::string::size_type position;
+            while (position = data_directory.find (subst_text),  position!=std::string::npos)
+              data_directory.replace (data_directory.begin()+position,
+                                      data_directory.begin()+position+subst_text.size(),
+                                      ASPECT_SOURCE_DIR);
+          }
+
+          data_file_name    = prm.get ("Data file name");
+
+          scale_factor      = prm.get_double ("Scale factor");
+          data_points[0]    = prm.get_double ("Number of x grid points");
```

Comment:
Duplication is the source of no ends of problems. I strongly vote for defining a file format that has a certain number of required (as opposed to optional) elements and then stick with it. If the number of points isn't allowed to change between input files, then test that in an assertion. But one could conceive of an extension where you do allow the number of points to change and do some clever interpolation somehow -- in which case the file format way will work, whereas the input parameter way will not.

But in the end as I say, let's not require the same information in different places.